src/detect-tls-cert.c

```diff
@@ -0,0 +1,142 @@
+
+#include "flow.h"
+#include "flow-util.h"
+#include "flow-var.h"
+
+#include "util-debug.h"
+#include "util-unittest.h"
+#include "util-spm.h"
+#include "util-print.h"
+
+#include "stream-tcp.h"
+
+#include "app-layer.h"
+#include "app-layer-ssl.h"
+
+#include "util-unittest.h"
+#include "util-unittest-helper.h"
+
+static int DetectTlsCertSetup(DetectEngineCtx *, Signature *, const char *);
+#ifdef UNITTESTS
+static void DetectTlsCertRegisterTests(void);
+#endif
+static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms,
+        Flow *_f, const uint8_t _flow_flags,
+        void *txv, const int list_id);
+static int g_tls_cert_buffer_id = 0;
+
+/**
+ * \brief Registration function for keyword: tls.cert
+ */
+void DetectTlsCertRegister(void)
+{
+    sigmatch_table[DETECT_AL_TLS_CERT].name = "tls.cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].desc = "content modifier to match the TLS certificate buffer";
+    sigmatch_table[DETECT_AL_TLS_CERT].url = DOC_URL DOC_VERSION "/rules/tls-keywords.html#tls.cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].Setup = DetectTlsCertSetup;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_AL_TLS_CERT].RegisterTests = DetectTlsCertRegisterTests;
+#endif
+    sigmatch_table[DETECT_AL_TLS_CERT].flags |= SIGMATCH_NOOPT;
+
+    DetectAppLayerInspectEngineRegister2("tls.cert", ALPROTO_TLS,
+            SIG_FLAG_TOCLIENT, TLS_STATE_CERT_READY,
+            DetectEngineInspectBufferGeneric, GetData);
+
+    DetectAppLayerMpmRegister2("tls.cert", SIG_FLAG_TOCLIENT, 2,
+            PrefilterGenericMpmRegister, GetData, ALPROTO_TLS,
+            TLS_STATE_CERT_READY);
+
+    DetectBufferTypeSetDescriptionByName("tls.cert", "TLS certificate");
+
+    g_tls_cert_buffer_id = DetectBufferTypeGetByName("tls.cert");
+}
+
+/**
+ * \brief This function setup the tls.cert modifier keyword
+ *
+ * \param de_ctx Pointer to the Detect Engine Context
+ * \param s      Pointer to the Signature to which the keyword belongs
+ * \param str    Should hold an empty string always
+ *
+ * \retval  0 On success
+ * \retval -1 On failure
+ */
+static int DetectTlsCertSetup(DetectEngineCtx *de_ctx, Signature *s,
+                              const char *str)
+{
+    if (DetectBufferSetActiveList(s, g_tls_cert_buffer_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_TLS) < 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *f,
+        const uint8_t flow_flags, void *txv, const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        const SSLState *ssl_state = (SSLState *)f->alstate;
+
+        if (TAILQ_EMPTY(&ssl_state->server_connp.certs)) {
+            return NULL;
+        }
+
+        const SSLCertsChain *cert = TAILQ_FIRST(&ssl_state->server_connp.certs);
```

Comment:
are we inspecting only one cert?

Reply:
Ok I see this is mentioned in the PR. Would it make sense to inspect all? Or would there be a case for a separate keyword for all certs or something?

Reply:
There's not really a good reason to only inspect one certificate. I'll change it to inspect the entire certificate chain :)

Reply:
Ok cool. This will complicate the inspection code somewhat, because you can now have multiple buffers per tx. I would suggest checking the dns query code which has the same issue.